htmresearch/algorithms/synapse_learning.py

```diff
@@ -0,0 +1,241 @@
+#!/usr/bin/env python
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2017, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""Implementations of various synapse learning rules"""
```

Comment:
This file basically controls how you sample from active inputs, and how many to sample. My first impression is that it is a lot of extra complexity (multiple classes, factory method, etc.) and a ton of extra code just for those options. And you still end up with an if-statement in line 209:
```
if self.maxSynapsesPerSegment != -1:
```

I don't see how this file is making the code simpler to understand - as far as I can tell it is taking something conceptually very simple and making it a lot more complex.